File: scripts/action_client_template.py
#! /usr/bin/env python
#coding:utf-8

import roslib; roslib.load_manifest('ros_template_programs')
import rospy
import actionlib
from ros_template_programs.msg import TimerAction
from ros_template_programs.msg import TimerGoal
from ros_template_programs.msg import TimerResult
from ros_template_programs.msg import TimerFeedback


class ActionClient():
    def __init__(self):
        self.client = actionlib.SimpleActionClient( 'timer_action_server', TimerAction)
        self.goal = TimerGoal()
        self.action_executing = False
        
    def send_goal(self):
        self.client.send_goal(
            self.goal,
            self.done_cb,
            self.active_cb,
            self.feedback_cb
        )
        return

    def done_cb(self, state, result):
        rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\nFinished!!\n' )
        rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\nTotal_time = %.2f\n', result.total_time )
        action_executing = False
        return
    
    def active_cb(self):
        rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\nGoal just went active...\n' )
        return
    
    def feedback_cb(self, feedback):
        rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\nGot Feedback of Progress to Goal\nelapsed_time = %.2f\n', feedback.elapsed_time )
        return

    def loop_call_server(self):
        loop_rate = rospy.Rate(1)
        rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\nWaiting For Server...\n' )
        self.client.wait_for_server()
        target_time = rospy.get_param( rospy.get_name() + '/target_time', 10.0)
        while not rospy.is_shutdown():
            if self.action_executing == False:
                self.goal.target_time = target_time;
                self.send_goal()
                rospy.loginfo('\n [ ' + rospy.get_name() + ' ]\nSet a new target_time = %.2f\n', self.goal.target_time )
                self.action_executing = True
            loop_rate.sleep()
        return

if __name__ == '__main__':
    try:
        rospy.init_node('action_client_template')
        
        node = ActionClient()
        node.loop_call_server()

        rospy.spin()

    except rospy.ROSInterruptException: pass

File: src/action_client_template.cpp
#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <actionlib/client/terminal_state.h>
#include <ros_template_programs/TimerAction.h>
#include <iostream>

namespace ros_template_programs {
    class ActionClient {
        private:
            ros::NodeHandle nh_;
            ros::NodeHandle pnh_;
            actionlib::SimpleActionClient<ros_template_programs::TimerAction> act_clt_;
            ros_template_programs::TimerGoal goal_;
            bool action_executing_;
            
            void sendGoal();
            void doneCb( const actionlib::SimpleClientGoalState& state, const ros_template_programs::TimerResultConstPtr& result );
            void activeCb( );
            void feedbackCb( const ros_template_programs::TimerFeedbackConstPtr& feedback );
        public:
            ActionClient( );
            void loopCallServer();
    };
}

void ros_template_programs::ActionClient::sendGoal() {
    act_clt_.sendGoal( goal_,
        boost::bind( &ros_template_programs::ActionClient::doneCb, this, _1, _2 ),
        boost::bind( &ros_template_programs::ActionClient::activeCb, this ),
        boost::bind( &ros_template_programs::ActionClient::feedbackCb, this, _1 ) );
}
void ros_template_programs::ActionClient::doneCb( 
    const actionlib::SimpleClientGoalState& state, 
    const ros_template_programs::TimerResultConstPtr& result ) {
    ROS_INFO( "[ %s ]\nFinished!!", ros::this_node::getName().c_str());
    ROS_INFO( "[ %s ]\nTotal_time = %.2f", ros::this_node::getName().c_str(), result->total_time );
    action_executing_ = false;
    ros::Duration(1.0).sleep();

}
void ros_template_programs::ActionClient::activeCb( ) {
    ROS_INFO( "[ %s ]\nGoal just went active...\n", ros::this_node::getName().c_str());
}
void ros_template_programs::ActionClient::feedbackCb( 
    const ros_template_programs::TimerFeedbackConstPtr& feedback ) {
    ROS_INFO( "[ %s ]\nGot Feedback of Progress to Goal\nelapsed_time = %.2f\n", ros::this_node::getName().c_str(), feedback->elapsed_time );
}

ros_template_programs::ActionClient::ActionClient( ) : nh_(), pnh_("~"),  act_clt_( "timer_action_server", true ) {
    ROS_INFO( "[ %s ] Waiting For Server...", ros::this_node::getName().c_str() );
    act_clt_.waitForServer();
    ROS_INFO("[ %s ] Connect to the action server", ros::this_node::getName().c_str());
    action_executing_ = false;
}

void ros_template_programs::ActionClient::loopCallServer( ) {
    ros::Rate loop_rate(1);
    double target_time = pnh_.param<double>("target_time", 10.0);
	while(ros::ok()) {
        if ( !action_executing_ ) {
            goal_.target_time = target_time;
            sendGoal();
            ROS_INFO( "[ %s ] Set a new target_time = %.2f\n", ros::this_node::getName().c_str(), target_time );
            action_executing_ = true;
        }
		ros::spinOnce();
		loop_rate.sleep();
	}
    return;
}

int main(int argc, char *argv[])  {
    ros::init(argc, argv, "action_client_template");
    ros_template_programs::ActionClient act_clt;
    act_clt.loopCallServer();
    ros::spin();
    return 0;
}